### PR TITLE
[GEP-22] Perform regular state backups for `Shoot`s running on unmanaged `Seed`s

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area scalability
/kind enhancement

**What this PR does / why we need it**:
As proposed in [GEP-22](https://github.com/gardener/gardener/blob/master/docs/proposals/22-improved-usage-of-shootstate-api.md), this PR introduces a new controller in `gardenlet` which is only activated in case `gardenlet` is responsible for an unmanaged `Seed` (i.e., one which is not backed by a `seedmanagement.gardener.cloud/v1alpha1.ManagedSeed` object).

By default, it will perform periodic backups (each `6h`) of the states of `Shoot`s running on its seed cluster. These backups are persisted in a `ShootState` resource in the same namespace as the `Shoot` in the garden cluster.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/8073

**Special notes for your reviewer**:
The next step in the implementation of GEP-22 is to drop the legacy [`shootstate-extensions`](https://github.com/gardener/gardener/blob/e130de2956448cd67deea3e97a4c3dd0d32d8494/pkg/gardenlet/controller/shootstate/extensions) and [`shootstate-secret`](https://github.com/gardener/gardener/blob/e130de2956448cd67deea3e97a4c3dd0d32d8494/pkg/gardenlet/controller/shootstate/secret) controllers.

/cc @timebertt 

**Release note**:

```feature operator
A new controller in `gardenlet` for periodically backing up the `ShootState` for `Shoot`s has been introduced. This controller is only activated when `gardenlet` is responsible for an unmanaged `Seed` (i.e., one not backed by a `ManagedSeed` object). By default, backups are taken roughly each `6h`.
```

### CODE REVIEW

File: docs/concepts/gardenlet.md [hunk 1]
```diff
@@ -440,6 +440,12 @@ A pod is considered stale when:
 - it was terminated with reason starting with `OutOf` (e.g., `OutOfCpu`).
 - it is stuck in termination (i.e., if its `deletionTimestamp` is more than `5m` ago).
 
+#### "State" Reconciler
+
+This reconciler periodically (default: every `6h`) performs backups of the state of `Shoot` clusters and persists them into `ShootState` resources into the same namespace as the `Shoot`s in the garden cluster.
+It is only started in case the `gardenlet` is responsible for an unmanaged `Seed`, i.e. a `Seed` which is not backed by a `seedmanagement.gardener.cloud/v1alpha1.ManagedSeed` object.
```
Comment (timebertt) [inline docs/concepts/gardenlet.md:446]: I was thinking about test coverage of the new scenarios coming with GEP-22.
In our e2e tests the cpm test shoots are always scheduled on a non-managed seed. With this, the controller would always be active and we wouldn't verify that migration also works without periodic backups.

Do you have plans for covering both cases already?

File: pkg/utils/gardener/shootstate/shootstate.go [hunk 2]
```diff
@@ -52,6 +55,7 @@ func Deploy(ctx context.Context, gardenClient, seedClient client.Client, shoot *
 	}
 
 	_, err = controllerutils.GetAndCreateOrStrategicMergePatch(ctx, gardenClient, shootState, func() error {
+		metav1.SetMetaDataAnnotation(&shootState.ObjectMeta, v1beta1constants.GardenerTimestamp, clock.Now().Format(time.RFC3339))
```
Comment (timebertt) [inline pkg/utils/gardener/shootstate/shootstate.go]:
```suggestion
		metav1.SetMetaDataAnnotation(&shootState.ObjectMeta, v1beta1constants.GardenerTimestamp, clock.Now().UTC().Format(time.RFC3339))
```

File: pkg/gardenlet/controller/shoot/state/reconciler.go [hunk 3]
```diff
@@ -55,5 +66,59 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
 
-	return reconcile.Result{}, nil
+	// if shoot got deleted or is no longer managed by this gardenlet (e.g., due to migration to another seed) then don't requeue
+	if shoot.DeletionTimestamp != nil || pointer.StringDeref(shoot.Spec.SeedName, "") != r.SeedName {
+		return reconcile.Result{}, nil
+	}
+
+	if !shootCreatedSuccessfully(shoot.Status) {
+		log.Info("Requeuing because shoot was not yet successfully created", "requeueAfter", RequeueWhenShootIsNotYetCreated)
+		return reconcile.Result{RequeueAfter: RequeueWhenShootIsNotYetCreated}, nil
```
Comment (timebertt) [inline pkg/gardenlet/controller/shoot/state/reconciler.go:76]: Instead of requeuing with a long duration, we could also adapt the update predicate to return true when the shoot just got successfully created.

File: pkg/gardenlet/controller/shoot/state/reconciler.go [hunk 4]
```diff
@@ -55,5 +66,59 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
+	if v, ok := shootState.Annotations[v1beta1constants.GardenerTimestamp]; ok {
+		var err error
+		lastBackup, err = time.Parse(time.RFC3339, v)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed parsing timestamp %q on ShootState object: %w", v, err)
+		}
+	}
+
+	if nextBackupDue := lastBackup.UTC().Add(r.Config.SyncPeriod.Duration); nextBackupDue.Before(r.Clock.Now().UTC()) {
+		log.Info("Performing periodic ShootState backup", "lastBackup", lastBackup.Round(time.Minute), "nextBackupDue", nextBackupDue.Round(time.Minute), "now", r.Clock.Now().UTC().Round(time.Minute))
```
Comment (timebertt) [inline pkg/gardenlet/controller/shoot/state/reconciler.go]: The `now` log field duplicates the timestamp log field, we shouldn't need it.

File: pkg/gardenlet/controller/shoot/state/reconciler.go [hunk 5]
```diff
@@ -55,5 +66,59 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
+
+	shootState := &gardencorev1beta1.ShootState{}
+	if err := r.GardenClient.Get(ctx, client.ObjectKeyFromObject(shoot), shootState); client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, fmt.Errorf("failed fetching ShootState %s: %w", client.ObjectKeyFromObject(shoot), err)
+	}
+
+	var lastBackup time.Time
+	if v, ok := shootState.Annotations[v1beta1constants.GardenerTimestamp]; ok {
+		var err error
+		lastBackup, err = time.Parse(time.RFC3339, v)
```
Comment (timebertt) [inline pkg/gardenlet/controller/shoot/state/reconciler.go:87]: Can we convert this back to UTC after parsing, so that the `lastBackup` log field is in the same time zone as the `nextBackupDue` log field?

File: test/integration/gardenlet/shoot/state/state_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	seedName         string
+
+	syncPeriod = 500 * time.Millisecond
+)
+
+var _ = BeforeSuite(func() {
+	logf.SetLogger(logger.MustNewZapLogger(logger.DebugLevel, logger.FormatJSON, zap.WriteTo(GinkgoWriter)))
+	log = logf.Log.WithName(testID)
+
+	features.RegisterFeatureGates()
```
Comment (timebertt) [inline test/integration/gardenlet/shoot/state/state_suite_test.go]: This shouldn't be needed here

File: test/integration/gardenlet/shoot/state/state_test.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+			&state.JitterDuration, time.Duration(0),
+		))
+
+		shoot = &gardencorev1beta1.Shoot{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "shoot-",
+				Namespace:    projectNamespace.Name,
+				Labels:       map[string]string{testID: testRunID},
+			},
+			Spec: gardencorev1beta1.ShootSpec{
```
Comment (timebertt) [inline test/integration/gardenlet/shoot/state/state_test.go:56]: Would be awesome to use a workerless shoot here :D

**Review by timebertt — COMMENTED** [review]
Awesome, nice PR!

File: docs/concepts/gardenlet.md (hunk 1 above)
Comment (rfranzke) [reply docs/concepts/gardenlet.md:446]: Good question :) What do you think about simply disabling this controller in the e2e migration tests? We already test its functionality with an integration test, so we don't need to run it in e2e.

File: pkg/gardenlet/controller/shoot/state/reconciler.go (hunk 3 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/shoot/state/reconciler.go:76]: Yes sure, I was thinking about it as well, but eventually decided against. We need this `if` condition in the reconciler anyways, and since periodic backups of the `ShootState`s are not time-critical, I decided to save the extra-effort of changing the watch predicates as suggested above. It's simply less code this way, and since we jitter anyways when computing the next backups, ~`10m` should still be good enough. If you insist, I can change it, but I don't see value in it.

File: docs/concepts/gardenlet.md (hunk 1 above)
Comment (timebertt) [reply docs/concepts/gardenlet.md:446]: Sounds good. I think performing integration test only for this controller is good enough, given that it is not required for actual migrations but rather supposed to be a fallback for edge cases.

File: pkg/gardenlet/controller/shoot/state/reconciler.go (hunk 3 above)
Comment (timebertt) [reply pkg/gardenlet/controller/shoot/state/reconciler.go:76]: I was also unsure if it was worth the effort. So agreed, let's keep it like this.

File: docs/concepts/gardenlet.md (hunk 1 above)
Comment (rfranzke) [reply docs/concepts/gardenlet.md:446]: Cool, then I'll do that in the next PR when I remove the old legacy controllers :)

**Review by timebertt — COMMENTED** [review]
Latest changes look good. Only https://github.com/gardener/gardener/pull/8112#discussion_r1236417469 left.

**Review by timebertt — APPROVED** [review]
